pkg/utils/kubernetes/referencegrant.go

```diff
@@ -0,0 +1,67 @@
+package kubernetes
+
+import (
+	"context"
+
+	"github.com/samber/lo"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	gatewayv1beta1 "sigs.k8s.io/gateway-api/apis/v1beta1"
+)
+
+// AllowedByReferenceGrants checks if the reference from the input `from` to the object(s)
+// in namespace `targetNamespace` with group, kind, name given in the input `to` is allowed by any ReferenceGrant.
+func AllowedByReferenceGrants(
+	ctx context.Context,
+	cl client.Client,
+	from gatewayv1beta1.ReferenceGrantFrom,
+	targetNamespace string,
+	to gatewayv1beta1.ReferenceGrantTo,
+) (bool, error) {
+	// Same namespace is always allowed.
+	if from.Namespace == gatewayv1beta1.Namespace(targetNamespace) {
+		return true, nil
+	}
+	referenceGrantList := gatewayv1beta1.ReferenceGrantList{}
+	namespacedClient := client.NewNamespacedClient(cl, targetNamespace)
+	err := namespacedClient.List(
+		ctx,
+		&referenceGrantList,
+	)
```

Comment:
I believe we could limit the scope of listing by indexing the reference grants by `from` entries. How about leaving a note here that we could improve it?

Reply:
Added a `// TODO` comment.